src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,63 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->compMethodReturnsMultiRegRetType())
+            {
```

Comment:
The suggested code snipped above has a bug. It doesn't handle the RetBuf return with the address of RetBuf returned in REG_INTRET on Windows(such case exists when the profiler hook is enabled).

Changing the genEmitGSCookieCheck() API is rather involved. It requires refactorings of code and it might/could affect even the Legacy codegen path. If this is to be done it should be done in a separate PR that focuses on this refactoring. Also the way we handle return types in different targets is still under discussion, so I'd defer this to later time.

PTAL